### ISSUE TITLE
EVT_CLOSE not firing in wx.Dialog

### ISSUE THREAD

**Issue by remdragon** [wxWidgets/Phoenix#672]
**Operating system**: Windows 10 Home
**wxPython version**: 4.0.0b2 msw (phoenix)
**Stock or custom build**: pip install --upgrade wxPython
**Python version**: Python 2.7.14 (v2.7.14:84471935ed, Sep 16 2017, 20:19:30) [MSC v.1500 32 bit (Intel)] on win32
**Stock or custom build**: (downloaded from python.org)

**Description of the problem**: The EVT_CLOSE event is ONLY being generated when the user clicks the "X" in the upper-right corner of the dialog. The issue is that it is *NOT* being generated when the user hits the ENTER or ESC key or when the user clicks the OK or Cancel buttons. According to the documentation url below, it should be generated in all these scenarios.

**Documentation URL**: https://wxpython.org/Phoenix/docs/html/wx.Dialog.html#events-events-emitted-by-this-class

```python
from __future__ import print_function

import wx

class Test_dialog ( wx.Dialog ):
	def __init__ ( self ):
		wx.Dialog.__init__ ( self, parent=None )
		self.Bind ( wx.EVT_CLOSE, self.on_close )
		self.SetSizerAndFit ( self.CreateButtonSizer ( wx.OK|wx.CANCEL ) )
	
	def on_close ( self, event ):
		print ( 'hello from on_close!' )
		self.Destroy()

if __name__=='__main__':
	app = wx.App ( False )
	dlg = Test_dialog()
	dlg.ShowModal()
```

**Comment by remdragon** [comment]
Tested on "3.0.2.0 msw (classic)" and the same problem exists there. Am I trying to use EVT_CLOSE wrong and/or misreading the documentation?

**Comment by RobinD42** [comment]
TL;DR: I don't think this is a bug.

The documentation does seem to imply that `EVT_CLOSE` would be sent in that case, however my recollection is that it has always been just for non-modal dialogs which are closed via the `Close` method. Modal dialogs are a little less straightforward.  They are normally closed by the ok or cancel button handlers (or your own button handlers if needed) calling `EndModal` with the ID that should be returned by the `ShowMocal` method. In fact, the default handler for `EVT_CLOSE` basically just does the same as if the button with ID `wx.ID_CANCEL` was clicked, so by default clicking the "X" will be the same as clicking Cancel, not the other way around.

Typical usage of modal dialogs is that control will wait in `ShowModal` until the dialog is accepted or cancelled, at which time `ShowModal` will return to the caller. At that point you do whatever you need to with the return value and/or the contents of the dialog, and then you destroy the dialog.

**Comment by remdragon** [comment]
Robin,

My normal method of doing this is by subclassing ShowModal(), but I took this route because a bit of paranoia about control lifetimes caused me to want to process the control values before closing the dialog.

In order to confirm, I just tweaked the test code and I can't seem to get EVT_CLOSE to work with a modeless dialog, either. However, I've rarely used this feature so I'm not sure if I'm initializing it correctly:

```python
from __future__ import print_function

import wx

class Test_dialog ( wx.Dialog ):
	def __init__ ( self ):
		wx.Dialog.__init__ ( self, parent=None )
		self.Bind ( wx.EVT_CLOSE, self.on_close )
		self.SetSizerAndFit ( self.CreateButtonSizer ( wx.OK|wx.CANCEL ) )
	
	def on_close ( self, event ):
		print ( 'hello from on_close!' )
		self.Destroy()

if __name__=='__main__':
	app = wx.App ( False )
	dlg = Test_dialog()
	app.SetTopWindow ( dlg )
	dlg.Show()
	app.MainLoop()
```

**Comment by RobinD42** [comment]
One way to process values before the dialog closes would be to use validators.  They can be assigned to the controls and their `Validate` method will be called when the dialog's ok button is clicked. The validators can display a message or update the UI in some way the value is not valid, and If any validators fail then the dialog will not be closed.

Another approach would be to intercept the `EVT_BUTTON` event for the ok button (or whatever your "Affermative ID" is) and do your checks there. If everything is okay then you can just call `EndModal` and the dialog will close and `ShowModal` will return to the caller.

**Comment by remdragon** [comment]
good to know about the validators, I've never used them before, but the reason I mentioned the modeless dialog is that it's behavior is matching that of the modal dialog which does not appear to be matching what the documentation says.

If it would help, I can reproduce this sample in C++ to see how the native wxWidgets library behaves.

**Comment by RobinD42** [comment]
There shouldn't be any C++/Python differences here.
